Add shields for Nova Scotia scenic drives

quincylvania opened the pull request:
Closes #911.

Creating a draft for sprite design feedback. Waiting for a tile refresh to review how these look on the map.

1ec5 commented on src/js/shield_defs.js:
```diff
@@ -4201,6 +4204,33 @@ export function loadShields() {
   // Ref-specific cases. Each entry should be documented in CONTRIBUTE.md
 
+  shields["CA:NS:S"].overrideByRef = {
+    BdOLSD: {
```
Do people or publications normally refer to the routes by these initialisms? If so, let’s add an explanation to CONTRIBUTING.md, in the section where we justify other unsignposted refs.

quincylvania replied:
I don't think so, I just made up the initialisms so we'd have something to key off of. I don't think we can key off the route name because I thiiiiink the name on the tile comes only from the way itself. The Merritt Parkway is the only one we do like that so far and it happens to have all way names match the parkway relation name.

This situation is kind of like the Great Lakes Circle Tour routes, but with a more items and less documentation. I imagine the problem will come up more often as we add additional networks of scenic routes.

I can add documentation for this.

1ec5 replied:
For #451, the New Zealand community apparently decided to give each route a distinct `network` subvalue in order to avoid polluting `ref` with unsignposted initialisms. However, this fragmentation makes less sense when the shield designs all follow a common motif, as in this case.

I think inevitably we’ll need the tiles to contain something beyond the `network=ref` syntax in `route_*` properties. If we could go back and change the format, I would’ve suggested something like `route:2:network` or `route_network_2`, which would’ve also logically allowed for `route_name_2` or even `route_wikidata_2` for these tricky cases.

1ec5 replied:
> This situation is kind of like the Great Lakes Circle Tour routes, but with a more items and less documentation.

The Great Lakes Circle Tour routes are known by the initialisms in `ref`, at least informally, so it isn’t a big deal if a non-shield-capable renderer displays the `ref` verbatim. I’d still reach out to the local community to get their opinion on this use of `ref`, in case there are more established initialisms for the routes.

1ec5 replied:
> Regarding the choice of ref on bespoke tourist routes, don't we have `overrideByName` as an option? Then we would be tied to the name, which should be beyond dispute. In any case, the choice of `ref` value on the route relation of an obscure tourist route seems like something we can just make up using reasonable conventions and change later if the community changes their mind down the line. In short, I'd favor pushing this forward either as-is or with a name-based definition.

@ZeLonewolf `overrideByName` was introduced for the Kentucky parkways. I agree that it would be an ideal solution… except that the `name` is just the way name, not the route relation’s name. For tourist routes, the road name isn’t reliably the route relation’s name, although it is in the case of the Cabot Trail.

quincylvania replied:
Yep, here's an illustrative example. Names like "Lighthouse Route" don't always appear in the data, so we're left to rely on a `ref` at the moment.

<img width="266" alt="Screenshot 2023-08-14 at 1 54 06 PM" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/c257d748-8ba9-4bae-8c5b-b50ebfd00db5">

ZeLonewolf replied:
Yep, makes sense to me - thanks for refreshing how that works. I built the damn logic for those `route_X` attributes and I can't even keep it straight...

1ec5 replied:
Let’s ticket out something about a `route_#_name` scheme in OpenMapTiles. Something similar would also help with Spain’s per-route colors in #654.

1ec5 replied:
As of #1032, each feature has a `route_#_name` property based on the route relation. The name gets plumbed through to the shield selection code, so you should be able to specify `overrideByName` without worrying about the roadway carrying a different name, and the relations no longer need to be tagged with initialisms as `ref`s.

<img src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/1231218/9ea42d6f-892c-43b9-9777-b0ad7a171bbd" width="214" alt="route_1_network: CA:NS:R; route_1_ref: 322; route_1_name: Route 322; route_2_network: CA:NS:S; route_2_name: Lighthouse Route">

1ec5 commented on icons/shield_ca_ns_s_md.svg:
I wonder if a few of these shields would be more legible if we don’t need to offset them to the left but can instead have them take up more space to the top and right.

quincylvania replied:
Sure, we could make the pictures slightly bigger by centering then and moving the water down. But personally I think I prefer sticking closer to the source designs. Though maybe seeing them on the map will affect this.

1ec5 replied:
Yeah, we’ll see how it turns out on the map. For the Ohio River Scenic Byway in #893, I made the shield 20&nbsp;pixels wide but taller than 20&nbsp;pixels, avoiding whitespace on either side, in order to maximize room for the shield’s details.

1ec5 replied:
> Regarding the choice of ref on bespoke tourist routes, don't we have `overrideByName` as an option? Then we would be tied to the name, which should be beyond dispute. In any case, the choice of `ref` value on the route relation of an obscure tourist route seems like something we can just make up using reasonable conventions and change later if the community changes their mind down the line. In short, I'd favor pushing this forward either as-is or with a name-based definition.

@ZeLonewolf `overrideByName` was introduced for the Kentucky parkways. I agree that it would be an ideal solution… except that the `name` is just the way name, not the route relation’s name. For tourist routes, the road name isn’t reliably the route relation’s name, although it is in the case of the Cabot Trail.

claysmalley replied:
The current shield for `network=CA:NS:H` is only 20px tall because we made it before relaxing the height guidelines for narrow shields. I think the designs introduced in this PR are good for now. If we want to enlarge them later, along with the Arterial shield, let's put that in a new issue.

ZeLonewolf commented:
Tiles are updated!

quincylvania commented:
In terms of size, I think the original 16x20 looks okay, perhaps too small:

![Screenshot 2023-08-07 at 2 48 17 PM](https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/ffd12741-7fa3-4579-8e11-2cb356fc066c)
![Screenshot 2023-08-07 at 2 49 28 PM](https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/ac943a4c-9af3-48fb-a2f2-2e93670fbd4e)

I tried 18x22 to keep even numbers and it seems a bit too big:

![Screenshot 2023-08-07 at 2 53 26 PM](https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/72064233-ab23-40c3-82e1-1a9018e1b38f)
![Screenshot 2023-08-07 at 2 53 38 PM](https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/06aaf6ca-34ab-4999-b0b7-f5e8c9ece8b4)

Ultimately I'd say 17x21 is a good compromise size as long as it doesn't cause rounding/aliasing issues.

![Screenshot 2023-08-07 at 3 09 53 PM](https://github.com/ZeLonewolf/openstreetmap-americana/assets/2046746/4e8fe9a3-24ad-4698-b1a9-b3add788311a)

ZeLonewolf commented:
Either the 17px or the 18px versions look pretty good to me in the screen shots.  I agree that the 16px one seems slightly undersized.

quincylvania commented:
I standardized on 17x21 px shields. I couldn't find any signage for the Fundy Shore Scenic Drive but otherwise this is complete and ready for review. FYI there was a `ref` mistake for the Sunrise Trail so that shield won't show up until a new tile update goes through but you can see it in the spritesheet.

1ec5 reviewed (commented):
These icons seem to hew to an 8-bit style, which happens to look great for these particular shields. I wonder if we should aim for a similar style in other shields, where we’re currently going for a very literal, realistic look and feel.

ZeLonewolf commented:
Regarding the choice of ref on bespoke tourist routes, don't we have `overrideByName` as an option?  Then we would be tied to the name, which should be beyond dispute. In any case, the choice of `ref` value on the route relation of an obscure tourist route seems like something we can just make up using reasonable conventions and change later if the community changes their mind down the line.  In short, I'd favor pushing this forward either as-is or with a name-based definition.

quincylvania commented:
This is ready for review btw. The discussion kind of got lost in the code comments but I think everything's been addressed.

ZeLonewolf commented on src/js/shield_defs.js:
```diff
@@ -4208,6 +4211,45 @@ export function loadShields() {
+    ET: {
+      spriteBlank: "shield_ca_ns_s_et",
+    },
+    FdLt: {
+      spriteBlank: "shield_ca_ns_s_fdlt",
+    },
```
This seems to have a case difference with https://www.openstreetmap.org/relation/11147926

quincylvania replied:
Thanks, fixed.

ZeLonewolf commented on src/js/shield_defs.js:
```diff
@@ -4208,6 +4211,45 @@ export function loadShields() {
+      spriteBlank: "shield_ca_ns_s_st",
+    },
+  };
+
```
FSSD appears to be missing:
https://www.openstreetmap.org/relation/11147926

quincylvania replied:
The FSSD is documented on Wikipedia but I couldn't find any signage for it.

ZeLonewolf reviewed (changes requested):
Looks good pending the fix for FDLT. We can either add FSSD here or in a separate ticket.

zekefarwell commented:
So if I understand correctly `ref=BdOLSD` on this route https://www.openstreetmap.org/relation/11051558 isn't a real identifier, but just made up to work around the fact that since no other field from a route relation is available in OpenMapTiles?  I wouldn't call myself a ground truth verifiability purist, but if that's the case this feels like a real stretch.  Seems to me like it would be better to work on improving OpenMapTiles with the desire for these type of scenic route shields as motivation.   I doubt these will be the last scenic routes folks will want to add.

quincylvania commented:
> So if I understand correctly ref=BdOLSD on this route https://www.openstreetmap.org/relation/11051558 isn't a real identifier, but just made up to work around the fact that no other field from a route relation is available in OpenMapTiles?

I wouldn't go so far as to say an initialism isn't a real identifier, but yes, these `ref` values are just unofficial workarounds (hacks). 

>  Seems to me like it would be better to work on improving OpenMapTiles with the desire for these type of scenic route shields as motivation. I doubt these will be the last scenic routes folks will want to add.

Agreed that something cleaner upstream would be nice, that there will be more cases in the future, and that this is the kind of issue Americana could push the envelope on.

But personally I think this is a "perfect is the enemy of good" scenario and would move forward as-is. We can improve the system later. But up to y'all.

zekefarwell commented:
> personally I think this is a "perfect is the enemy of good" scenario and would move forward as-is. We can improve the system later.

I also would consider it such a scenario if this was a fully self contained project.  However, we are just one part of the OSM ecosystem, and I don't think we want to be seen as a project that supports questionable tagging practices.  The stakes are quite low in this case, but I think it would be best to wait for a proper solution in OpenMapTiles.

ZeLonewolf commented:
I think openmaptiles/openmaptiles#1576 is not that far off, if the team is on board with that approach.  While testing it, I did find a pretty major bug in openmaptiles that's causing a massive reduction in shields from z11 and lower.  So my plan is to fix that issue and that push 1576 along.  Since it's been a very long time since the last openmaptiles release. I'm willing to be we're not that far off from a proper solution for ref-less routes.

ZeLonewolf commented:
Okay, the prerequisite is done.

I could use some design help in figuring out what the implementation for openmaptiles/openmaptiles#1576 should actually do to support these use cases.

ZeLonewolf commented:
Looks like the deploy preview is hit by actions/first-interaction#10.  I'll plan to do some research as to how that can be resolved.

ZeLonewolf commented:
The CI issues have been resolved.  The schema update remain unresolved.

1ec5 reviewed (commented):
From my perspective, the artwork in this PR is ready to ship, so the only thing that needs to change is migrating from `overrideByRef` to `overrideByName`.

quincylvania commented:
I've changed this to use `overrideByName` so I think it should be ready to go. I removed the made-up `ref` tags in this changeset: https://www.openstreetmap.org/changeset/150189553

claysmalley reviewed (approved):
[Looks good!](https://preview.ourmap.us/pr/912/#map=10.08/46.1729/-60.6214)

![image](https://github.com/ZeLonewolf/openstreetmap-americana/assets/1732117/359cba21-04f7-4ffa-ae0a-5131f9199b96)

ZeLonewolf commented:
Thanks for sticking this one out!  Looks awesome!

quincylvania commented:
Thanks for running a project that improves tooling and sees things through even if it takes eight months!
